File: Lista de Exercícios/Python/Escola.py
# Questão 7
class Escola:
    def __init__(self, nome):
        self.nome = nome
        self.professores = []

    def adicionar_professor(self, professor):
        if professor not in self.professores:
            self.professores.append(professor)
            professor.adicionar_escola(self)

    def remover_professor(self, professor):
        if professor in self.professores:
            self.professores.remove(professor)
            professor.remover_escola(self)

    def listar_professores(self):
        return [professor.nome for professor in self.professores]

class Professor:
    def __init__(self, nome):
        self.nome = nome
        self.escolas = []

    def adicionar_escola(self, escola):
        if escola not in self.escolas:
            self.escolas.append(escola)

    def remover_escola(self, escola):
        if escola in self.escolas:
            self.escolas.remove(escola)

    def listar_escolas(self):
        return [escola.nome for escola in self.escolas]

# Exemplo de uso
if __name__ == "__main__":
    escola1 = Escola("Escola A")
    escola2 = Escola("Escola B")

    prof1 = Professor("João")
    prof2 = Professor("Maria")

    escola1.adicionar_professor(prof1)
    escola1.adicionar_professor(prof2)
    escola2.adicionar_professor(prof1)

    print(f"Professores da {escola1.nome}: {escola1.listar_professores()}")
    print(f"Professores da {escola2.nome}: {escola2.listar_professores()}")
    print(f"Escolas onde {prof1.nome} leciona: {prof1.listar_escolas()}")
    print(f"Escolas onde {prof2.nome} leciona: {prof2.listar_escolas()}")

File: Lista de Exercícios/Golang/Escola.go
// Questão 7
package main

import (
	"fmt"
)

// Escola representa uma instituição de ensino
type Escola struct {
	Nome        string
	Professores []*Professor
}

// AdicionarProfessor adiciona um professor à escola
func (e *Escola) AdicionarProfessor(professor *Professor) {
	for _, p := range e.Professores {
		if p == professor {
			return
		}
	}
	e.Professores = append(e.Professores, professor)
	professor.AdicionarEscola(e)
}

// RemoverProfessor remove um professor da escola
func (e *Escola) RemoverProfessor(professor *Professor) {
	for i, p := range e.Professores {
		if p == professor {
			e.Professores = append(e.Professores[:i], e.Professores[i+1:]...)
			professor.RemoverEscola(e)
			return
		}
	}
}

// ListarProfessores retorna uma lista com os nomes dos professores
func (e *Escola) ListarProfessores() []string {
	nomes := make([]string, len(e.Professores))
	for i, p := range e.Professores {
		nomes[i] = p.Nome
	}
	return nomes
}

// Professor representa um professor
type Professor struct {
	Nome    string
	Escolas []*Escola
}

// AdicionarEscola adiciona uma escola ao professor
func (p *Professor) AdicionarEscola(escola *Escola) {
	for _, e := range p.Escolas {
		if e == escola {
			return
		}
	}
	p.Escolas = append(p.Escolas, escola)
}

// RemoverEscola remove uma escola do professor
func (p *Professor) RemoverEscola(escola *Escola) {
	for i, e := range p.Escolas {
		if e == escola {
			p.Escolas = append(p.Escolas[:i], p.Escolas[i+1:]...)
			return
		}
	}
}

// ListarEscolas retorna uma lista com os nomes das escolas
func (p *Professor) ListarEscolas() []string {
	nomes := make([]string, len(p.Escolas))
	for i, e := range p.Escolas {
		nomes[i] = e.Nome
	}
	return nomes
}

func main() {
	escola1 := &Escola{Nome: "Escola A"}
	escola2 := &Escola{Nome: "Escola B"}

	prof1 := &Professor{Nome: "João"}
	prof2 := &Professor{Nome: "Maria"}

	escola1.AdicionarProfessor(prof1)
	escola1.AdicionarProfessor(prof2)
	escola2.AdicionarProfessor(prof1)

	fmt.Printf("Professores da %s: %v\n", escola1.Nome, escola1.ListarProfessores())
	fmt.Printf("Professores da %s: %v\n", escola2.Nome, escola2.ListarProfessores())
	fmt.Printf("Escolas onde %s leciona: %v\n", prof1.Nome, prof1.ListarEscolas())
	fmt.Printf("Escolas onde %s leciona: %v\n", prof2.Nome, prof2.ListarEscolas())
}

